File: internal/config/validate.go
package config

import (
	"fmt"
	"regexp"
	"strings"
	"time"
)

// ValidationError collects multiple validation problems.
type ValidationError struct {
	Errors []string
}

func (ve *ValidationError) Error() string {
	return strings.Join(ve.Errors, "\n")
}

func (ve *ValidationError) addf(format string, args ...any) {
	ve.Errors = append(ve.Errors, fmt.Sprintf(format, args...))
}

func (ve *ValidationError) hasErrors() bool {
	return len(ve.Errors) > 0
}

var (
	appNameRe = regexp.MustCompile(`^[a-z0-9-]+$`)
	domainRe  = regexp.MustCompile(`^[a-z0-9]([a-z0-9-]*[a-z0-9])?(\.[a-z0-9]([a-z0-9-]*[a-z0-9])?)*$`)
)

// Validate checks the entire config and returns all errors found.
func Validate(cfg *Config) error {
	ve := &ValidationError{}

	// Poll interval must be a valid Go duration.
	if cfg.PollInterval != "" {
		if _, err := time.ParseDuration(cfg.PollInterval); err != nil {
			ve.addf("poll_interval: invalid duration %q", cfg.PollInterval)
		}
	}

	// Tunnel.
	if cfg.Tunnel != nil {
		if cfg.Tunnel.Provider != "cloudflare" {
			ve.addf("tunnel: provider must be 'cloudflare', got %q", cfg.Tunnel.Provider)
		}
	}

	// Notification channels.
	for name, ch := range cfg.Notifications {
		if !ValidNotifyDrivers[ch.Driver] {
			ve.addf("notification '%s': driver must be telegram, got %q", name, ch.Driver)
		}
	}

	// GitHub providers.
	if cfg.GitHub != nil {
		for name, p := range cfg.GitHub.Providers {
			prefix := fmt.Sprintf("github.providers '%s'", name)
			if !appNameRe.MatchString(name) {
				ve.addf("%s: name must match [a-z0-9-]+", prefix)
			}
			if p.Type != "key" && p.Type != "app" {
				ve.addf("%s: type must be 'key' or 'app', got %q", prefix, p.Type)
			}
			if p.Type == "app" && p.AppID <= 0 {
				ve.addf("%s: app_id is required for type 'app'", prefix)
			}
		}
	}

	// Apps.
	for name, app := range cfg.Apps {
		validateApp(ve, name, &app, cfg.GitHub, cfg.Notifications)
	}

	if ve.hasErrors() {
		return ve
	}
	return nil
}

func validateApp(ve *ValidationError, name string, app *App, github *GitHubConfig, notifications map[string]NotificationChannel) {
	prefix := fmt.Sprintf("app '%s'", name)

	// App name format.
	if !appNameRe.MatchString(name) {
		ve.addf("%s: name must match [a-z0-9-]+", prefix)
	}

	// Required: repo.
	if app.Repo == "" {
		ve.addf("%s: repo is required", prefix)
	}

	// Provider must reference an existing github provider (if set).
	if app.Provider != "" && app.Repo != "local" {
		found := false
		if github != nil {
			if _, ok := github.Providers[app.Provider]; ok {
				found = true
			}
		}
		if !found {
			ve.addf("%s: provider %q not found in github.providers", prefix, app.Provider)
		}
	}

	// Required: at least one domain.
	if len(app.Domains) == 0 {
		ve.addf("%s: at least one domain is required", prefix)
	}

	// Domains.
	for i, d := range app.Domains {
		dprefix := fmt.Sprintf("%s: domain '%s'", prefix, d.Host)
		if d.Host == "" {
			ve.addf("%s: domain[%d]: host is required", prefix, i)
		} else if !domainRe.MatchString(d.Host) {
			ve.addf("%s: invalid hostname", dprefix)
		}
		if d.Port < 1 || d.Port > 65535 {
			ve.addf("%s: invalid port %d", dprefix, d.Port)
		}
		if d.Ingress != "" && !ValidIngressValues[d.Ingress] {
			ve.addf("%s: ingress must be 'direct' or 'cloudflare-tunnel', got %q", dprefix, d.Ingress)
		}
	}

	// Health checks.
	for _, h := range app.Health {
		if !strings.HasPrefix(h.Path, "/") {
			ve.addf("%s: health check path must start with '/', got %q", prefix, h.Path)
		}
		if h.Port < 1 || h.Port > 65535 {
			ve.addf("%s: health check invalid port %d", prefix, h.Port)
		}
	}

	// Warmup duration.
	if app.Warmup != "" {
		if _, err := time.ParseDuration(app.Warmup); err != nil {
			ve.addf("%s: warmup: invalid duration %q", prefix, app.Warmup)
		}
	}

	// Notify: each referenced channel must exist in notifications.
	for _, ch := range app.Notify {
		if _, ok := notifications[ch]; !ok {
			ve.addf("%s: notify references undefined channel %q", prefix, ch)
		}
	}
}

File: modules/config/validate.ts
import { ConfigError } from './errors.ts'
import type { Config } from './schema.ts'

const APP_NAME_RE = /^[a-z0-9][a-z0-9-]*$/
const DOMAIN_RE = /^[a-z0-9]([a-z0-9-]*[a-z0-9])?(\.[a-z0-9]([a-z0-9-]*[a-z0-9])?)*$/
// GitHub slug: `owner/name` — both segments [A-Za-z0-9._-], no `..`, no slashes.
const GITHUB_SLUG_RE = /^[A-Za-z0-9][A-Za-z0-9._-]*\/[A-Za-z0-9][A-Za-z0-9._-]*$/
const EXTERNAL_REPO_PREFIXES = ['file://', 'http://', 'https://', 'ssh://', 'git://']

/**
 * Returns an error message if `repo` is not a supported shape, else `null`.
 * Accepted: `"local"`, `""`, an absolute path, a scheme URL, `git@host:path`,
 * or an `owner/name` GitHub slug. Rejects anything containing `..` segments
 * or embedded slashes beyond the one in the GitHub slug — those would let a
 * maliciously-crafted config escape `$JIB_ROOT/repos/` via `repoPath`.
 */
export function validateRepo(repo: string): string | null {
  if (repo === '' || repo === 'local') return null
  if (repo.includes('..')) return 'contains ".." path segment'
  if (repo.startsWith('/')) return null
  if (repo.startsWith('git@') && /^git@[^:\s]+:[^\s]+$/.test(repo)) return null
  for (const prefix of EXTERNAL_REPO_PREFIXES) {
    if (repo.startsWith(prefix)) return null
  }
  if (GITHUB_SLUG_RE.test(repo)) return null
  return 'must be "local", "owner/name", an absolute path, or a file://, http(s)://, ssh://, git://, or git@host: URL'
}

/**
 * Parses a duration string like `5m`, `30s`, `1h`, `1h30m`, `1.5h`, `0s`.
 * Returns milliseconds, or `null` on parse failure. Supported units: `s`, `m`,
 * `h`. Non-negative only — `-5s`, empty strings, unit-less numbers (`5`), and
 * any unknown unit all return `null`. Narrower than Go's `time.ParseDuration`
 * (no `ns`/`us`/`ms`) but covers every value jib's config actually uses.
 */
export function parseDuration(s: string): number | null {
  if (!s) return null
  const units: Record<string, number> = { s: 1000, m: 60_000, h: 3_600_000 }
  const re = /(\d+(?:\.\d+)?)([smh])/g
  let total = 0
  let matched = 0
  for (const m of s.matchAll(re)) {
    const [, n, unit] = m
    const mult = unit ? units[unit] : undefined
    if (!mult || n === undefined) return null
    total += Number.parseFloat(n) * mult
    matched += m[0].length
  }
  return matched === s.length && matched > 0 ? total : null
}

/** Runs config-level checks that zod can't express. Throws `ConfigError`. */
export function validate(cfg: Config): void {
  const errs: string[] = []

  if (parseDuration(cfg.poll_interval) === null) {
    errs.push(`poll_interval: invalid duration "${cfg.poll_interval}"`)
  }

  const sourceNames = new Set(Object.keys(cfg.sources))
  for (const name of sourceNames) {
    if (!APP_NAME_RE.test(name)) {
      errs.push(`sources '${name}': name must match [a-z0-9-]+`)
    }
    const source = cfg.sources[name]
    if (source?.driver === 'github' && source.type === 'app' && source.app_id === undefined) {
      errs.push(`sources '${name}': app_id is required for driver github type app`)
    }
  }

  let needsTunnel = false
  for (const [name, app] of Object.entries(cfg.apps)) {
    if (!APP_NAME_RE.test(name)) errs.push(`app '${name}': name must match [a-z0-9-]+`)
    const repoErr = validateRepo(app.repo)
    if (repoErr) errs.push(`app '${name}': repo "${app.repo}" ${repoErr}`)
    if (app.source && app.repo !== 'local' && !sourceNames.has(app.source)) {
      errs.push(`app '${name}': source "${app.source}" not found in sources`)
    }
    for (const d of app.domains) {
      if (d.host !== d.host.toLowerCase() || !DOMAIN_RE.test(d.host)) {
        errs.push(`app '${name}': invalid hostname "${d.host}"`)
      }
      if (d.ingress === 'cloudflare-tunnel') needsTunnel = true
    }
    for (const h of app.health ?? []) {
      if (!h.path.startsWith('/')) {
        errs.push(`app '${name}': health check path must start with '/'`)
      }
    }
  }

  if (needsTunnel && !cfg.tunnel) {
    errs.push('tunnel: config required when any domain uses cloudflare-tunnel ingress')
  }

  if (errs.length > 0) throw new ConfigError(errs.join('\n'))
}
